File: backtracking/046_permutations.cpp
```
class Solution {
public:
    vector<vector<int>> res;
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> tmp;
        backtrack(nums, tmp);
        return res;
    }
    
    void backtrack(vector<int>& nums, vector<int>& tmp) {
        if (nums.size() == tmp.size()) {
            res.push_back(tmp);
            return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (std::find(tmp.begin(), tmp.end(), nums[i]) != tmp.end()) continue;
            tmp.push_back(nums[i]);
            backtrack(nums, tmp);
            tmp.pop_back();
        }
    }
};
```

File: backtracking/046_permutations.java
// 046
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList();
        List<Integer> tmp = new ArrayList();
        permute(nums, tmp, res);
        return res;
    }
    
    public void permute(int[] nums, List<Integer> tmp, List<List<Integer>> res) {
        if (nums.length == tmp.size()) {
            res.add(new ArrayList(tmp));
            return;
        }
        
        int n = nums.length;
        for (int i = 0; i < n; ++i) {
            if (tmp.contains(nums[i])) continue;
            tmp.add(nums[i]);
            permute(nums, tmp, res);
            tmp.remove(tmp.size() - 1);
        }
    }
}
```

